lib/puppetserver/ca/generate_action.rb

```diff
@@ -0,0 +1,249 @@
+
+        Puppetserver::Utils::FileUtilities.ensure_dir(puppet.settings[:cadir])
+
+        future_files = [ { location: puppet.settings[:cacert], content: [int_cert, root_cert]},
+                         { location: puppet.settings[:cakey], content: int_key },
+                         { location: puppet.settings[:rootkey], content: root_key },
+                         { location: puppet.settings[:cacrl], content: [int_crl, root_crl]}
+                       ]
+
+        future_files.each do | file|
+          @logger.warn "#{file[:location]} exists, overwriting" if File.exist?(file[:location])
+          Puppetserver::Utils::FileUtilities.write_file(file[:location], file[:content], 0640)
+        end
+      end
+
+      def create_private_key(length)
+        OpenSSL::PKey::RSA.new(length.to_i)
+      end
+
+      def self_signed_ca(key, name, ca_ttl)
+        cert = OpenSSL::X509::Certificate.new
+
+        cert.public_key = key.public_key
+        cert.subject = OpenSSL::X509::Name.new([["CN", name]])
+        cert.issuer = cert.subject
+        cert.version = 2
+        cert.serial = 1
+
+        # Make the certificate valid as of yesterday, because so many people's
+        # clocks are out of sync.  This gives one more day of validity than people
+        # might expect, but is better than making every person who has a messed up
+        # clock fail, and better than having every cert we generate expire a day
+        # before the user expected it to when they asked for "one year".
+        cert.not_before = Time.now - (60*60*24)
+        cert.not_after  = Time.now + ca_ttl
+
+        ef = extension_factory_for(cert, cert)
+        CA_EXTENSIONS.each do |ext|
+          extension = ef.create_extension(*ext)
+          cert.add_extension(extension)
+        end
+
+        cert.sign(key, DEFAULT_SIGNING_DIGEST)
+
+        cert
+      end
+
+      def extension_factory_for(ca, cert = nil)
+        ef = OpenSSL::X509::ExtensionFactory.new
+        ef.issuer_certificate  = ca
+        ef.subject_certificate = cert if cert
+
+        ef
+      end
+
+      def create_crl_for(ca_cert, ca_key, ca_ttl)
+        crl = OpenSSL::X509::CRL.new
+        crl.version = 1
+        crl.issuer = ca_cert.subject
+
+        ef = extension_factory_for(ca_cert)
+        crl.add_extension(
+          ef.create_extension(["authorityKeyIdentifier", "keyid:always", false]))
+        crl.add_extension(
+          OpenSSL::X509::Extension.new("crlNumber", OpenSSL::ASN1::Integer(0)))
+
+        crl.last_update = Time.now - 1
+        crl.next_update = Time.now + ca_ttl
+        crl.sign(ca_key, DEFAULT_SIGNING_DIGEST)
+
+        crl
+      end
+
+      def create_csr(key, name)
+        csr = OpenSSL::X509::Request.new
+
+        csr.public_key = key.public_key
+        csr.subject = OpenSSL::X509::Name.new([["CN", name]])
+        csr.version = 2
+        csr.sign(key, DEFAULT_SIGNING_DIGEST)
+
+        csr
+      end
+
+      def sign(ca_key, ca_cert, csr, ca_ttl)
```

Comment:
Since we hard code the serial number to be two it should only be used to sign the intermediate CA. Since it's not a generic sign method we should make that fact more explicit in the method name.